protocols/imap/src/main/java/org/apache/james/imap/processor/base/SelectedMailboxImpl.java

```diff
@@ -396,6 +392,28 @@ private void mailboxEvent(MailboxEvent mailboxEvent) {
         }
     }
 
+    private void updateApplicableFlags(FlagsUpdated messageEvent) {
+        Flags accumulator = Optional.ofNullable(applicableFlags.get())
+            .map(Flags::new)
+            .orElse(new Flags());
+        int initialSize = accumulator.getUserFlags().length;
+        List<UpdatedFlags> updatedFlags = messageEvent.getUpdatedFlags();
+
+        for (UpdatedFlags flag : updatedFlags) {
+            accumulator.add(flag.getNewFlags());
+        }
+
+        // \RECENT is not a applicable flag in imap so remove it
+        // from the list
+        accumulator.remove(Flag.RECENT);
+
+        if (initialSize < accumulator.getUserFlags().length) {
+            applicableFlagsChanged = true;
```

Comment:
this property is not protected from concurrency and it makes the whole class subject to dirty-read as you update it separately from `applicableFlags`

Reply:
I merely extracted the existing code, that I know is far from perfect (and bug free) in order to ship a minimal change-set addressing the original issue I came for.

If we do plan to test it, detect additional bugs, solve them and improve code quality here, then we need to allocate more work, I suggest adding a comment on the existing JIRA ticket and keeping it open.

Would you agree with this?

Reply:
> I merely extracted the existing code, that I know is far from perfect (and bug free) in order to ship a minimal change-set addressing the original issue I came for.
> 
> If we do plan to test it, detect additional bugs, solve them and improve code quality here, then we need to allocate more work, I suggest adding a comment on the existing JIRA ticket and keeping it open.
> 
> Would you agree with this?

As far as I understand this PR is about fixing a race condition. You found some tests reproducing some of them and fixed them.

However, I find it relevant to spot other race-conditions, I'm not nick-picking on a random change here or there, right?

So what about protecting properly from the cases of concurrency that we know will occur?